man/mcp-check.1
.\" Copyright (c) 2019-2021 Miki Hermann & Gernot Salzer
.TH mcp-check 1 "2024-08-19" "1.04" "MCP System"
.
.SH NAME
mcp-check - checking the accuracy of a formula
.
.SH SYNOPSIS
.B mcp-check
.RI [\| "OPTION" "\|]\|.\|.\|."
.
.SH DESCRIPTION
.PP
Takes a formula produced by the MCP core and checks its accuracy
against a set of Boolean vectors prefixed by a group identifier. The
accuracy of the checked formula is reported in terms of true positives
(tp), true negatives (tn), false positive (fp), false negatives (fn),
sensitivity (tp / (tp + fn)), miss rate (fn / (fn + tp)), specificity
(tn / (tn + fp)), and precision (tp / (tp + fp)).
.
.SH OPTIONS
.
.TP
\fB\-i\fR, \fB\-\-input\fI input-file
Boolean vectors prefixed by group identifiers used to check the
formulas accuracy.
.IP
Default: STDIN.
.
.TP
\fB\-o\fR, \fB\-\-output\fI output-file
Report of the checking run.
.IP
Default: STDOUT.
.
.TP
\fB\-l\fR, \fB\-\-formula\fI formula-file
File containing the formula produced by MCP core. The
\fIformula-file\fR has the name \fIformula-prefix_G.log\fR if it has
been produced by the option "\fB\-\-formula \fIformula-prefix\fR" for
some group \fIG\fR.
.
.TP
\fB\-\-hdr\fR, \fB\-\-header\fI header-file
Internal file containing the variables names produced by MCP transformation.
This parameter is rarely explicitly specified.
The header file is produced by the module \fImcp-trans(1)\fR.
.IP
Default: \fIbasename\fR of \fIinput-file\fR with \fI.hdr\fR suffix.
.
.TP
\fB\-\-print\fR clause | implication | mix | dimacs
Printing format of the used formula.
.IP
clause:
.br
Print each clause as a disjunction of positive and negative literals.
.br
Example: (-x0 + x1) * (x0 + -x1).
.IP
implication:
.br
Print each clause as an implication.
.br
Example: (x0 -> x1) * (x1 -> x0).
.IP
mix:
.br
If there are only literals of one polarity in a clause, then print as
"clause", otherwise as "implication".
.IP
dimacs:
.br
Print the formula in DIMACS format, one clause per line (implies
\fB\-\-offset\fR greater than 0).
.IP
Default: mix.
.
.TP
\fB\-m\fR, \fB\-\-matrix\fR undefined | hide | peek | section | show
Controls how input and output matrices are printed.
.IP
undefined:
.br
Printing of the input matrices is undefined; their status will be
deduced later from the arity and size of the input. The matrices after
computing a minimal section will NOT be printed.
.IP
hide:
No matrices will be printed.
.IP
peek:
.br
The input matrices will be printed, the matrices after minimal section
will NOT be printed.
.IP
section:
.br
The input matrices will NOT be printed, the matrices after minimal
section will be printed.
.IP
show:
.br
Both the input matrices and matrices after computing the minimal section will be
printed.
.IP
Default: undefined.
.
.PP
.
.SH SEE ALSO
mcp-guess(1),
mcp-trans(1),
mcp-split(1),
mcp-seq(1),
mcp-mpi(1),
mcp-pthread(1),
mcp-hybrid(1),
mcp-sparse(1),
mcp-chk2tst(1),
mcp-predict(1)
.
.SH BUGS
There are certanly some.
.
.SH AUTHORS
Miki Hermann <hermann@lix.polytechnique.fr>
.br
Gernot Salzer <gernot.salzer@tuwien.ac.at>


changelog.txt
1.04
- fixed a bug in mcp-check.cpp
- added LaTeX output
- mcp-split introduced
- post_prod introduced
- write formula to a file (DIMACS form)

1.03c
 - added print option DIMACS
 - corrected sorting bug when '--cooking raw'
 - corrected bug when '--strategy exact' and '--closure Horn'
 - index of group_of_matrix: int -> string
 - variables can be named in the matrix
 - transformation in C++
 - failure to open input or output file results in exit and not in exception
 - added options 'peek' and 'section' for printing matrices
 - when variables have own names, literals made readable
 - bug correction: 'continue' instead of 'return' when empty formula
   in sequential for bijunctive 
 - added mixed printing of formulas
 - added resolution with binary clause which contracts the target clause
   (l0 + l1) * (l0 + -l1 + c) => (l0 + l1) * (l0 + c)
 - added mcp-cnf producting exact CNF formula


src-seine/mcp-check.cpp
/**************************************************************************
 *                                                                        *
 *                                                                        *
 *	         Multiple Classification Project (MCP)                    *
 *                                                                        *
 *	Author:   Miki Hermann                                            *
 *	e-mail:   hermann@lix.polytechnique.fr                            *
 *	Address:  LIX (CNRS UMR 7161), Ecole Polytechnique, France        *
 *                                                                        *
 *	Author: Gernot Salzer                                             *
 *	e-mail: gernot.salzer@tuwien.ac.at                                *
 *	Address: Technische Universitaet Wien, Vienna, Austria            *
 *                                                                        *
 *	Version: all                                                      *
 *      File:    mcp-check.cpp                                            *
 *                                                                        *
 *      Copyright (c) 2019 - 2025                                         *
 *                                                                        *
 * Given a  formula and  a set  of boolean  vectors, check  which vectors *
 * satisfy  and which  falsify the  given formula.  Report precision  and *
 * recall (https://en.wikipedia.org/wiki/Precision_and_recall).           *
 *                                                                        *
 **************************************************************************/

#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include "mcp-matrix+formula.hpp"

using namespace std;

const string STDIN  = "STDIN";
const string STDOUT = "STDOUT";

string input  = STDIN;
string output = STDOUT;
string headerput;
string formula_input;
ifstream infile;
ifstream form_in;
ifstream headerfile;
ofstream outfile;

Formula formula;

vector<size_t> names;
// string suffix;
// int arity;
// int nvars;
// int offset;

int tp = 0;		// true positive
int tn = 0;		// true negative
int fp = 0;		// false postitive
int fn = 0;		// false negative
double tpr = RSNTNL;	// true positive rate aka sensitivity aka recall
double tnr = RSNTNL;	// true negative rate aka specificity aka selectivity
double ppv = RSNTNL;	// positive predictive value aka precision
double npv = RSNTNL;	// negative predictive value
double fnr = RSNTNL;	// false negative rate aka miss rate
double fpr = RSNTNL;	// false positive rate aka fall-out
double fdr = RSNTNL;	// false discovery rate
double forate = RSNTNL;	// false omission rate
double pt = RSNTNL;	// prevalence treshhold
double csi = RSNTNL;	// critical success index aka threat score
double acc = RSNTNL;	// accuracy
double ba = RSNTNL;	// balanced accuracy
double f1score = RSNTNL;// F1 score


//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

void read_arg (int argc, char *argv[]) {	// reads the input parameters
  int argument = 1;
  while (argument < argc) {
    string arg = argv[argument];
    if (arg == "--input"
	|| arg == "-i") {
      if (argument < argc-1) {
	input = argv[++argument];
      } else
	cerr << "+++ no input file selected" << endl;
    } else if (arg == "--hdr"
	       || arg == "--header") {
      if (argument < argc-1) {
	headerput = argv[++argument];
      } else
	cerr << "+++ no header file selected, revert to default" << endl;
    } else if (arg == "--output"
    	       || arg == "-o") {
      if (argument < argc-1) {
	output = argv[++argument];
      } else
	cerr << "+++ no output file selected, revert to default" << endl;
    } else if (arg == "--formula"
	       || arg == "--logic"
	       || arg == "--log"
	       || arg == "-l") {
      if (argument < argc-1) {
	formula_input = argv[++argument];
      } else
	cerr << "+++ no formula file prefix selected, revert to default" << endl;
    } else if (arg == "--pr"
	       || arg == "--print") {
      if (argument < argc-1) {
	string prt = argv[++argument];
	if (prt == "clause"
	    || prt == "clausal"
	    || prt == "cl"
	    || prt == "c") {
	  print_val = pCLAUSE;
	} else if (prt == "implication"
		   || prt == "impl"
		   || prt == "imp"
		   || prt == "im"
		   || prt == "i") {
	  print_val = pIMPL;
	} else if (prt == "mix"
		   || prt == "mixed"
		   || prt == "m") {
	  print_val = pMIX;
	} else if (prt == "dimacs"
		   || prt == "DIMACS") {
	  print_val = pDIMACS;
	} else
	  cerr <<  "+++ unknown print option " << prt << endl;
      } else
	cerr << "+++ no print option selected, revert to default" << endl;
    } else if (arg == "--matrix"
	       || arg == "--mtx"
	       || arg == "-m") {
      if (argument < argc-1) {
	string mtx = argv[++argument];
	if (mtx == "yes"
	    || mtx == "y"
	    || mtx == "show") {
	  display = ySHOW;
	} else if (mtx == "peek") {
	  display = yPEEK;
	} else if (mtx == "no"
		   || mtx == "n"
		   || mtx == "hide") {
	  display = yHIDE;
	} else if (mtx == "undefined"
		   || mtx == "undef"
		   || mtx == "u") {
	  display = yUNDEF;
	} else
	  cerr << "+++ unknown matrix print option " << mtx << endl;
      } else
	cerr << "+++ no matrix printing selected, revert to default" << endl;
    } else
      cerr << "+++ unknown option " << arg << endl;
    ++argument;
  }
}

void adjust_and_open () {
  form_in.open(formula_input);
  if (form_in.is_open())
    cin.rdbuf(form_in.rdbuf());
  else {
    cerr << "+++ Cannot open formula input file " << formula_input << endl;
    exit(2);
  }

  if (input != STDIN && headerput.empty()) {
    string::size_type pos = input.rfind('.');
    headerput = (pos == string::npos ? input : input.substr(0, pos)) + ".hdr";
  }

  if (output != STDOUT) {
    outfile.open(output);
    if (outfile.is_open())
      cout.rdbuf(outfile.rdbuf());
    else {
      cerr << "+++ Cannot open output file " << output << endl;
      exit(2);
    }
  }

  if (print_val == pVOID)
    print_val = pMIX;
}

void print_arg () {
  cout << "@@@ Parameters:" << endl;
  cout << "@@@ ===========" << endl;
  cout << "@@@ version       = " << version << endl;
  cout << "@@@ input         = " << input << endl;
  cout << "@@@ header        = " << headerput << endl;
  cout << "@@@ output        = " << output << endl;
  cout << "@@@ var. offset   = " << offset << endl;
  cout << "@@@ print matrix  = " << display_strg[display]
       << (display == yUNDEF ? " (will be changed)" : "") << endl;
  cout << "@@@ print formula = " << print_strg[print_val] << endl;
  cout << "@@@ formula input = " << (formula_input.empty() ? "none" : formula_input) << endl;
  cout << endl;
}

void read_header () {
  if (headerput.empty())
    varswitch = false;
  else {
    headerfile.open(headerput);
    if (! headerfile.is_open()) {
      cerr << "+++ Cannot open header file " << headerput << endl
	   << "... Continue with fake variable names" << endl;
      varswitch = false;
      return;
    }

    cout << "+++ Own names for variables" << endl;
    varswitch = true;

    string line;
    while(getline(headerfile, line)) {
      const vector<string> &hdl = split(line, ":");
      varnames.push_back(hdl);
    }
    // arity = varnames.size();

    headerfile.close();
  }
}

void read_matrix (Group_of_Matrix &matrix) {
  if (input != STDIN) {
    infile.open(input);
    if (infile.is_open())
      cin.rdbuf(infile.rdbuf());
    else {
      cerr << "+++ Cannot open input file " << input << endl;
      exit(2);
    }
  }

  // matrix read instructions
  // maybe to be changed and replaced with the one in mcp-seq
  // reads the input matrices

  string line;
  string group;
  int numline = 0;
  while (getline(cin, line)) {
    numline++;
    istringstream nums(line);
    nums >> group;
    Row temp;
    int number;
    while (nums >> number)
      temp.push_back(number);
    if (arity != temp.size())
      cout << "*** arity discrepancy on line " << numline << endl;

    matrix[group].push_back(temp);
  }
  
  if (input != STDIN)
    infile.close();

  if (display == yUNDEF) {
    display = (numline * arity > MTXLIMIT) ? yHIDE : yPEEK;
    cout << "@@@ print matrix  = " << display_strg[display]
       << " (redefined)" << endl;
  }
}

// copied from mcp-seq, but slightly changed (use gmtx.size() instead numline)
// need to do something to have only one copy
void print_matrix (const Group_of_Matrix &matrix) {
  // prints the matrices
  cout << "+++ Arity = " << arity << endl;
  for (auto group = matrix.cbegin(); group != matrix.cend(); ++group) {
    cout << "+++ Group " << group->first;
    grps.push_back(group->first);
    Matrix gmtx = group->second;
    cout << " [" << gmtx.size() << "]:" << endl;
    if (display == yPEEK || display == ySHOW)
      cout << gmtx << endl;
  }
  sort(grps.begin(), grps.end());
  cout << "+++ Number of groups = " << grps.size() << endl;

  cout << endl;
  cout << "+++ Satisfying: " << suffix << endl;
  cout << "+++ Falsifying:";
  for (auto group : grps)
    if (group != suffix)
      cout << " " << group;
  cout << endl << endl;
}

void print_formula (const vector<size_t> &names, const Formula &formula) {
  string strg_fm = formula2string(names, formula);
  cout << "+++ Formula [" << formula.size() << "] =" << endl;
  cout << strg_fm << endl;
}

void sat_test (const Group_of_Matrix &matrix, const Formula &formula) {
  for (auto group = matrix.begin(); group != matrix.end(); ++group) {
    Matrix gmtx = group->second;
    for (Row row : gmtx)
      if (group->first == suffix) {		
	// must satisfy
	if (sat_formula(row, formula))
	  tp++;
	else
	  fn++;
      } else {
	// must falsify
	if (sat_formula(row, formula))
	  fp++;
	else
	  tn++;
      }
  }
  if (tp+fn != 0) {
    tpr = (1.0 * tp) / (1.0*tp + 1.0*fn);
    fnr = 1.0 - tpr;
  }
  if (tn+fp != 0) {
    tnr = (1.0 * tn) / (1.0*tn + 1.0*fp);
    fpr = 1.0 * fp / (1.0*fp + 1.0*tn);
  }
  if (tp+fp != 0)
    ppv = (1.0 * tp) / (1.0*tp + 1.0*fp);
  if (tp+tn+fp+fn != 0)
    acc = 1.0 * (tp + tn) / (1.0*tp + 1.0*tn + 1.0*fp + 1.0*fn);
  if (tpr+ppv != 0.0)
    f1score = (1.0 * tp) / (1.0*tp + 0.5*fp + 0.5*fn);
}

inline size_t max (const size_t a, const size_t b) {
  return a >= b ? a : b;
}

void print_result () {
  size_t maxnum = 0;
  maxnum = max(maxnum, tp);
  maxnum = max(maxnum, tn);
  maxnum = max(maxnum, fp);
  maxnum = max(maxnum, fn);

  const size_t perclen = 6;
  const size_t numlen = max(to_string(maxnum).length(), perclen)+1;

  cout << endl;
  cout << "+++ Statistics:" << endl;
  cout << "    ===========" << endl;
  cout << "+++ true  positive (tp)  = " << setw(numlen) << right << tp << endl;
  cout << "+++ true  negative (tn)  = " << setw(numlen) << right << tn << endl;
  cout << "+++ false positive (fp)  = " << setw(numlen) << right << fp << endl;
  cout << "+++ false negative (fn)  = " << setw(numlen) << right << fn << endl;

  cout << "+++ sensitivity    (tpr) = ";
  if (tpr < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << tpr * 100.0 << " %";
  cout << right << "\t [tp / (tp + fn)]" << endl;

  cout << "+++ miss rate      (fnr) = ";
  if (fnr < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << fnr * 100.0 << " %";
  cout << right << "\t [fn / (fn + tp)]" << endl;

  cout << "+++ fall-out       (fnr) = ";
  if (fpr < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << fpr * 100.0 << " %";
  cout << right << "\t [fp / (fp + tn)]" << endl;

  cout << "+++ specificity    (tnr) = ";
  if (tnr < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << tnr * 100.0 << " %";
  cout << right << "\t [tn / (tn + fp)]" << endl;

  cout << "+++ precision      (ppv) = ";
  if (ppv < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << ppv * 100.0 << " %";
  cout << right << "\t [tp / (tp + fp)]" << endl;

  cout << "+++ accuracy       (acc) = ";
  if (acc < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << acc * 100.0 << " %";
  cout << right << "\t [(tp + tn) / (tp + tn + fp + fn)]" << endl;

  cout << "+++ F_1 score      (F_1) = ";
  if (f1score < 0.0)
    cout << " ---.--  ";
  else
    cout << right << setw(perclen+1) << setprecision(2) << fixed << showpoint
	 << f1score * 100.0 << " %";
  cout << right << "\t [tp / (tp + 0.5 * (fp +fn))]" << endl;

  form_in.close();
  if (output != STDOUT)
    outfile.close();
}

//////////////////////////////////////////////////////////////////////////////

int main(int argc, char **argv)
{
  version += "check";
  cerr << "+++ version = " << version << endl;

  read_arg(argc, argv);
  adjust_and_open();
  print_arg();
  read_formula(names, formula);
  read_header();
  read_matrix(group_of_matrix);
  print_matrix(group_of_matrix);
  print_formula(names, formula);
  sat_test(group_of_matrix, formula);
  print_result();
}

//////////////////////////////////////////////////////////////////////////////
